library/Backends/llvm/src/Juvix/Backends/LLVM/Parameterization.hs

```diff
@@ -8,24 +9,81 @@ module Juvix.Backends.LLVM.Parameterization
   )
 where
 
+import qualified Data.List.NonEmpty as NonEmpty
 import Juvix.Backends.LLVM.Primitive
 import qualified Juvix.Core.Application as App
 import qualified Juvix.Core.Base.Types as Core
 import qualified Juvix.Core.IR.Evaluator as IR
 import qualified Juvix.Core.Parameterisation as Param
 import Juvix.Library
+import Juvix.Library.Usage
 import qualified LLVM.AST.Type as LLVM
 
 instance Param.CanApply PrimTy where
   -- TODO: Needs to implement apply
   arity = arityTy
 
 instance Param.CanApply (PrimVal ext) where
-  -- TODO: Needs to implement apply
+  type Arg (PrimVal ext) = App.Arg' ext (Param.PrimType PrimTy) RawPrimVal
+
+  -- We need to define @pureArg@ as we overwrite the default @Arg@ definition.
+  pureArg = undefined
+
   arity val = case val of
     App.Cont {} -> App.numLeft val
     App.Return {} -> arityRaw $ App.retTerm val
 
+  -- The following implementation follows the eval/apply method for curried
+  -- function application. A description of this can be found in 'How to make a
+  -- fast curry: push/enter vs eval/apply' by Simon Marlow and Simon Peyton
+  -- Jones.
+  -- Given a function, and a non-empty list of arguments, we try to apply the
+  -- arguments to the function. The function is of type 'PrimVal''/'Return'',
+  -- so either a continuation or a fully evaluated term.
+  apply fun args = do
+    let (fun', args', funArity') = toTakes fun -- 'args'' are part of continuation fun'
+        argLen = lengthN args -- Nr. of free arguments.
+        argsAll = foldr NonEmpty.cons args args' -- List of all arguments.
+    case argLen `compare` funArity' of
+      -- If there are not enough arguments to apply, return a continuation.
+      LT ->
+        Right $
+          App.Cont
+            { App.fun = fun',
+              App.args = toList argsAll,
+              App.numLeft = funArity' - argLen
+            }
+      -- If there are exactly enough arguments to apply, do so.
+      -- In case there aren't any arguments, return a continuation.
+      EQ
+        | Just returns <- traverse App.argToReturn argsAll ->
+          applyProper (takeToReturn fun) returns |> first Param.Extra
+        | otherwise ->
+          Right $
+            App.Cont
+              { App.fun = fun',
+                App.args = toList argsAll,
+                App.numLeft = 0
+              }
+      -- If there are too many arguments to apply, raise an error.
+      GT -> Left $ Param.ExtraArguments fun args
+    where
+      toTakes App.Cont {App.fun, App.args, App.numLeft} = (fun, args, numLeft)
+      toTakes App.Return {App.retType, App.retTerm} = (fun, [], arityRaw retTerm)
+        where
+          fun =
+            App.Take
+              { App.usage = Omega,
```

Comment:
Is this because the `usage` in `App.Take` isn't actually used?  Or is it used but `Omega` is always correct?  Or is usage-tracking just not implemented here yet?

Reply:
This code was incomplete and therefore removed from the branch for now. I will need to add something back later.